libtenzir/builtins/operators/sort.cpp

```diff
@@ -309,8 +312,188 @@ class plugin final : public virtual operator_plugin<sort_operator> {
+public:
+  sort_operator2() = default;
+
+  explicit sort_operator2(std::vector<sort_expression> sort_exprs)
+    : sort_exprs_{std::move(sort_exprs)} {
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.sort";
+  }
+
+  auto
+  operator()(generator<table_slice> input,
+             operator_control_plane& ctrl) const -> generator<table_slice> {
+    auto events = std::vector<table_slice>{};
+    auto indices = std::vector<sort_index>{};
+    auto sort_keys = std::vector<sort_key>{};
+    sort_keys.reserve(sort_exprs_.size());
+    for (const auto& sort_expr : sort_exprs_) {
+      sort_keys.emplace_back().reverse = sort_expr.reverse;
+    }
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      const auto length = detail::narrow<int64_t>(slice.rows());
+      indices.reserve(indices.size() + slice.rows());
+      for (int64_t i = 0; i < length; ++i) {
+        indices.push_back({
+          .slice = events.size(),
+          .event = i,
+        });
+      }
+      for (size_t i = 0; i < sort_exprs_.size(); ++i) {
+        sort_keys[i].chunks.push_back(
+          eval(sort_exprs_[i].expr, slice, ctrl.diagnostics()));
+      }
+      events.push_back(std::move(slice));
+    }
+    if (indices.empty()) {
+      co_return;
+    }
+    // TODO: If all chunks for a sort key evaluate to the same type, then we can
+    // choose a faster path where we do not need to evaluate the sort key's type
+    // for each row individually. That may be significantly faster.
+    std::ranges::sort(indices, [&](const sort_index& lhs,
+                                   const sort_index& rhs) {
+      for (const auto& sort_key : sort_keys) {
+        const auto& lhs_key = sort_key.chunks[lhs.slice];
+        const auto& rhs_key = sort_key.chunks[rhs.slice];
+        const auto lhs_null = lhs_key.array->IsNull(lhs.event);
+        const auto rhs_null = rhs_key.array->IsNull(rhs.event);
+        if (lhs_null and rhs_null) {
+          continue;
+        }
+        if (lhs_null or rhs_null) {
+          // Nulls last, independent of sort order.
+          return rhs_null;
+        }
+        const auto& lhs_value
+          = value_at(lhs_key.type, *lhs_key.array, lhs.event);
+        const auto& rhs_value
+          = value_at(rhs_key.type, *rhs_key.array, rhs.event);
+        // TODO: Implement this directly on data and data_view
+        const auto cmp = detail::overload{
+          [](const concepts::integer auto& l, const concepts::integer auto& r) {
+            return std::cmp_less(l, r);
+          },
+          [](const concepts::number auto& l, const concepts::number auto& r) {
+            return l < r;
+          },
+          [&](const auto& l, const auto& r) {
+            if constexpr (std::same_as<decltype(l), decltype(r)>) {
+              return l < r;
+            }
+            return lhs_value.index() < rhs_value.index();
+          }};
```

Comment:
The thing that would be missing here is to use the same logic recursively for records fields and list elements. We could skip this for now though, it's more of an edge case anyway. But it should probably be included in the TODO comment or in the story.

Reply:
We *tried* to do this, but it leads to rather difficult issues with overloads nested in lambdas in std::function.

Given that we aren't sure whether the compression logic for nested records even is the behavior we want, we just accept that the comparison numeric values in nested records is different for now.